tools/undelete-blocks/main.go

```diff
@@ -0,0 +1,471 @@
+
+	"github.com/grafana/mimir/pkg/storage/tsdb/block"
+	"github.com/grafana/mimir/pkg/util/objtools"
+)
+
+type config struct {
+	bucketConfig    objtools.BucketConfig
+	includeTenants  flagext.StringSliceCSV
+	excludeTenants  flagext.StringSliceCSV
+	inputFile       string
+	inputFileFormat string
+	dryRun          bool
+}
+
+func (c *config) registerFlags(f *flag.FlagSet) {
+	c.bucketConfig.RegisterFlags(f)
+	f.Var(&c.includeTenants, "include-tenants", "If not empty, only blocks for these tenants are considered.")
+	f.Var(&c.excludeTenants, "exclude-tenants", "Blocks for these tenants are not considered.")
+	f.StringVar(&c.inputFile, "input-file", "", "If set, the blocks to undelete are read from the provided file.")
+	f.StringVar(&c.inputFileFormat, "input-file-format", "json", "The format of the input file. Accepted values: json and lines.")
+	f.BoolVar(&c.dryRun, "dry-run", false, "If true, no writes/deletes are performed and are instead only logged.")
+}
+
+func main() {
+	cfg := config{}
+	cfg.registerFlags(flag.CommandLine)
+
+	// Parse CLI arguments.
+	if err := flagext.ParseFlagsWithoutArguments(flag.CommandLine); err != nil {
+		fmt.Fprintln(os.Stderr, err.Error())
+		os.Exit(1)
+	}
+
+	if err := cfg.bucketConfig.Validate(); err != nil {
+		fmt.Fprintln(os.Stderr, err.Error())
+		os.Exit(1)
+	}
+
+	ctx, cancel := signal.NotifyContext(context.Background(), syscall.SIGINT, syscall.SIGTERM)
+	defer cancel()
+
+	err := run(ctx, cfg)
+	if err != nil {
+		fmt.Fprintln(os.Stderr, err.Error())
+		os.Exit(1)
+	}
+}
+
+func run(ctx context.Context, cfg config) error {
+	bucket, err := cfg.bucketConfig.ToBucket(ctx)
+	if err != nil {
+		return errors.Wrap(err, "failed to create bucket")
+	}
+
+	blocks, err := getBlocks(ctx, cfg, bucket)
+	if err != nil {
+		return errors.Wrap(err, "failed to get blocks")
+	}
+
+	undeleteBlocks(ctx, bucket, blocks, cfg.dryRun)
+	return nil
+}
+
+type tenantFilter func(string) bool
+
+func newTenantFilter(cfg config) tenantFilter {
+	includeTenants := map[string]struct{}{}
+	for _, u := range cfg.includeTenants {
+		includeTenants[u] = struct{}{}
+	}
+	excludeTenants := map[string]struct{}{}
+	for _, u := range cfg.excludeTenants {
+		excludeTenants[u] = struct{}{}
+	}
+
+	return func(tenantID string) bool {
+		if len(includeTenants) > 0 {
+			if _, ok := includeTenants[tenantID]; !ok {
+				return false
+			}
+		}
+		_, ok := excludeTenants[tenantID]
+		return !ok
+	}
+}
+
+func getBlocks(ctx context.Context, cfg config, bucket objtools.Bucket) (map[string][]ulid.ULID, error) {
+	tenantFilter := newTenantFilter(cfg)
+	if cfg.inputFile != "" {
+		switch strings.ToLower(cfg.inputFileFormat) {
+		case "json":
+			return getBlocksFromJSONFile(cfg.inputFile, tenantFilter)
+		case "lines":
+			return getBlocksFromLinesFile(cfg.inputFile, tenantFilter)
+		default:
+			return nil, errors.Errorf("unrecognized input file format: %s", cfg.inputFileFormat)
+		}
+	}
+	return getBlocksFromListing(ctx, bucket, tenantFilter)
+}
```

Comment:
Let's get blocks from stdin, by reading paths to the blocks. I don't think we need to support JSON vs lines, or input file. (stdin works fine).

Reply:
> Let's get blocks from stdin, by reading paths to the blocks.

Sure, to do this I've added `--blocks-from`, which gives the choices `json`, `lines`, and `listing`. If `--input-file` is not supplied with `json` or `lines` it now defaults to reading from stdin. The net effect is you can say 
```bash
echo "tenant/01GVC9CD6KGMNPVB47X9S208NF" | ./undelete-blocks --blocks-from lines
```

> I don't think we need to support JSON vs lines

We've talked about this externally, but for posterity the reason I want JSON support was to provide a pathway for selecting on multiple block attributes with something like `jq`.